lib/mix/lib/mix/tasks/test.ex

```diff
@@ -659,69 +665,115 @@ defmodule Mix.Tasks.Test do
 
     warn_files != [] && warn_misnamed_test_files(warn_files)
 
-    try do
-      Enum.each(test_paths, &require_test_helper(shell, &1))
-      # test_opts always wins because those are given via args
-      ExUnit.configure(ex_unit_opts |> merge_helper_opts() |> Keyword.merge(test_opts))
-      CT.require_and_run(matched_test_files, test_paths, test_elixirc_options, opts)
-    catch
-      kind, reason ->
-        # Also mark the whole suite as failed
-        file = Keyword.fetch!(opts, :failures_manifest_path)
-        ExUnit.Filters.fail_all!(file)
-        :erlang.raise(kind, reason, __STACKTRACE__)
+    if dry_run? do
+      do_dry_run(matched_test_files)
+    else
+      do_wet_run(
+        shell,
+        cover,
+        test_paths,
+        files,
+        matched_test_files,
+        test_opts,
+        test_elixirc_options,
+        ex_unit_opts,
+        warnings_as_errors?,
+        exit_status,
+        opts
+      )
+    end
+  end
+
+  defp do_dry_run(matched_test_files) do
+    if matched_test_files == [] do
+      Mix.shell().info("""
+      -- DRY RUN --
+      No tests would run
+      """)
     else
-      {:ok, %{excluded: excluded, failures: failures, warnings?: warnings?, total: total}} ->
-        Mix.shell(shell)
-        cover && cover.()
+      Mix.shell().info("""
+      -- DRY RUN --
+      The following test files would be run:
+
+      #{Enum.join(matched_test_files, "\n")}
+      """)
+    end
+  end
 
-        cond do
-          warnings_as_errors? and warnings? and failures == 0 ->
-            message =
-              "\nERROR! Test suite aborted after successful execution due to warnings while using the --warnings-as-errors option"
+  defp do_wet_run(
```

Comment:
This is partially meant as a joke, but I am open to better name/organization options!